packages/scm/src/browser/decorations/scm-decorations-service.ts

```diff
@@ -15,64 +15,82 @@
 // *****************************************************************************
 
 import { injectable, inject } from '@theia/core/shared/inversify';
-import { ResourceProvider } from '@theia/core';
-import { DirtyDiffDecorator } from '../dirty-diff/dirty-diff-decorator';
+import { DisposableCollection, Emitter, Event, ResourceProvider } from '@theia/core';
+import { DirtyDiffDecorator, DirtyDiffUpdate } from '../dirty-diff/dirty-diff-decorator';
 import { DiffComputer } from '../dirty-diff/diff-computer';
 import { ContentLines } from '../dirty-diff/content-lines';
-import { EditorManager, TextEditor } from '@theia/editor/lib/browser';
+import { EditorManager, EditorWidget, TextEditor } from '@theia/editor/lib/browser';
 import { ScmService } from '../scm-service';
 
+import throttle = require('@theia/core/shared/lodash.throttle');
+
 @injectable()
 export class ScmDecorationsService {
-    private readonly diffComputer: DiffComputer;
-    private dirtyState: boolean = true;
+    private readonly diffComputer = new DiffComputer();
+
+    protected readonly onDirtyDiffUpdateEmitter = new Emitter<DirtyDiffUpdate>();
+    readonly onDirtyDiffUpdate: Event<DirtyDiffUpdate> = this.onDirtyDiffUpdateEmitter.event;
 
-    constructor(@inject(DirtyDiffDecorator) protected readonly decorator: DirtyDiffDecorator,
+    constructor(
+        @inject(DirtyDiffDecorator) protected readonly decorator: DirtyDiffDecorator,
         @inject(ScmService) protected readonly scmService: ScmService,
         @inject(EditorManager) protected readonly editorManager: EditorManager,
-        @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider) {
-        this.diffComputer = new DiffComputer();
-        this.editorManager.onCreated(async editor => this.applyEditorDecorations(editor.editor));
-        this.scmService.onDidAddRepository(repository => repository.provider.onDidChange(() => {
-            const editor = this.editorManager.currentEditor;
-            if (editor) {
-                if (this.dirtyState) {
-                    this.applyEditorDecorations(editor.editor);
-                    this.dirtyState = false;
-                } else {
-                    /** onDidChange event might be called several times one after another, so need to prevent repeated events. */
-                    setTimeout(() => {
-                        this.dirtyState = true;
-                    }, 500);
-                }
+        @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider
+    ) {
+        const updateTasks = new Map<EditorWidget, () => void>();
+        this.editorManager.onCreated(editorWidget => {
+            const { editor } = editorWidget;
+            if (editor.uri.scheme !== 'file') {
```

Comment:
I guess it was by analogy with:

https://github.com/eclipse-theia/theia/blob/96685e37b5450a599370355ccffd4062cf1b72ab/packages/git/src/browser/dirty-diff/dirty-diff-manager.ts#L72-L77

I agree that this restriction might be a bit too strong, although I don't have a specific case in mind when it would not hold.

However, I think that we still need to ensure that at least diff editors are ignored here, and probably some other kinds of editors too.

What would be the best way to handle it in your opinion?

Reply:
Since the whole thing does not work for anything other than git anyway, the check is probably correct. But if we have the same concern in the `DirtyDiffManager`, why not centralize the method there?

Reply:
> But if we have the same concern in the `DirtyDiffManager`, why not centralize the method there?

I don't quite understand it, could you please explain what exactly I need to do here? :-)

Reply:
In both locations, we're trying to determine whether to show the "dirty diff" for a given editor. Why not extract the check to to a method `shouldDisplayDirtyDiff(EditorWidget)`? This way way make sure the code is the same in both places and we document what the check is supposed mean.

Reply:
Thank you for the explanation.

I think that it would make sense to extract the check to a protected `shouldDisplayDirtyDiff(EditorWidget)` method in each of the `DirtyDiffManager` and `ScmDecorationsService` classes. It would allow clients to override those methods if necessary and would also document what the check is supposed to mean.

However, I'm not quite sure about the necessity of a shared implementation, i.e. about *make sure the code is the same in both places* part. While `DirtyDiffManager` is specific to git, `ScmDecorationsService` is clearly not, so the implementations of  the `shouldDisplayDirtyDiff(EditorWidget)` methods in those classes could potentially be quite different, although they would be the same for now. Besides, It is also unclear to me which class could contain such a shared implementation, if needed.

Reply:
I'd also propose `supportsDirtyDiff(EditorWidget)` instead of `shouldDisplayDirtyDiff(EditorWidget)`, if you don't mind.

Reply:
Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/e2857119646e3e2fc7ca11116838eccfd780f09a.

I have added `TextEditor.shouldDisplayDirtyDiff()` method. Its implementation in `MonacoEditor` returns `true`, while its implementation in `MonacoDiffEditor` returns `false`. This largely corresponds to the `isCodeEditor()` check in `DirtyDiffWorkbenchController` in [VS Code](https://github.com/microsoft/vscode/blob/d8a01ed97e89cad9a7ef5b380449ae822898c684/src/vs/workbench/contrib/scm/browser/dirtydiffDecorator.ts#L1652-L1654).

This method is called from protected `supportsDirtyDiff(TextEditor)` methods, which have been added to `DirtyDiffManager` and `ScmDecorationsService`, as proposed above.

Reply:
@tsmaeder Could you please take a look at this?